dipy/align/imaffine.py

```diff
@@ -0,0 +1,777 @@
+        points_on_moving = points_on_moving[..., :self.dim]
+        moving_32 = self.moving.astype(np.float32)
+        self.moving_vals, inside = self.interp_method(moving_32,
+                                                      points_on_moving)
+        self.moving_vals = np.array(self.moving_vals, dtype=np.float64)
+
+        # Update the joint and marginal intensity distributions
+        self.update_pdfs_sparse(self.static_vals, self.moving_vals)
+
+        # Compute the gradient of the joint PDF w.r.t. parameters
+        if update_gradient:
+            # Compute the gradient of the moving image at the current transform
+            mgrad, inside = vf.sparse_gradient(self.moving.astype(np.float32),
+                                               sp_to_moving,
+                                               self.moving_spacing,
+                                               self.samples)
+            self.update_gradient_sparse(xopt, self.transform, self.static_vals,
+                                        self.moving_vals,
+                                        self.samples[..., :self.dim],
+                                        mgrad)
+
+        # Evaluate the mutual information and its gradient
+        # The results are in self.metric_val and self.metric_grad
+        # ready to be returned from 'distance' and 'gradient'
+        self.update_mi_metric(update_gradient)
+
+    def distance(self, xopt):
+        r""" Numeric value of the negative Mutual Information
+
+        We need to change the sign so we can use standard minimization
+        algorithms.
+
+        Parameters
+        ----------
+        xopt : array, shape (n,)
+            the parameter vector of the transform currently used by the metric
+            (the transform name is provided when self.setup is called), n is
+            the number of parameters of the transform
+
+        Returns
+        -------
+        val : float
+            the negative mutual information of the input images after warping
+            the moving image by the currently set transform with `xopt`
+            parameters
+        """
+        if self.samples is None:
+            self._update_dense(xopt, False)
+        else:
+            self._update_sparse(xopt, False)
+        return -1 * self.metric_val
+
+    def gradient(self, xopt):
+        r""" Numeric value of the metric's gradient at the given parameters
+
+        Parameters
+        ----------
+        xopt : array, shape (n,)
+            the parameter vector of the transform currently used by the metric
+            (the transform name is provided when self.setup is called), n is
+            the number of parameters of the transform
+
+        Returns
+        -------
+        grad : array, shape(n,)
+            the gradient of the negative Mutual Information
+        """
+        if self.samples is None:
+            self._update_dense(xopt, True)
+        else:
+            self._update_dense(xopt, True)
```

Comment:
Nice catch again!, now we have only one `update`, so this error won't happen again..